### PR TITLE
Refactor upgrade for control plane components

### PR DESCRIPTION
**Pull request by lentzi90**
- Use Ansible modules k8s and k8s_info instead of shell for interaction
with the API.
- Use relevant Ansible modules instead of shell for file operations
(edit/copy)
- Use label selectors instead of regex search for relevant k8s resources
- Add Ansible filter plugin for removing autogenerated fields (e.g.
resourceVersion and uid)

### CODE REVIEW

**Comment by lentzi90** [comment]
/hold
This PR is using the `CAPI_VERSION` variable, which is currently set to `v1alpha4`. I'll need to hard code to `v1alpha3` until #688 is fixed.

**Review by macaptain — APPROVED** [review]
Great work. The filter plugin looks really nice, great way of solving that problem, and I appreciate the unit tests for it too.

I checked the build time for the upgrade tests, and there's no change. The extra build output we get about the BareMetalHosts will help with debugging problems as we'll be able to see the full state of the k8s resource on failures.

There are some v1alpha3 references here. I am not sure what we need to do to prepare for v1alpha4 -> v1alpha5 testing, but I think this is outside the scope of your refactor.

**Review by furkatgofurov7 — COMMENTED** [review]
This is great to see, good job!

**Comment by furkatgofurov7** [comment]
> There are some v1alpha3 references here. I am not sure what we need to do to prepare for v1alpha4 -> v1alpha5 testing, but I think this is outside the scope of your refactor.

I think we have to make sure that if this PR lands before that API uplift work, we need to update those references most probably in #693 after rebasing.

**Comment by lentzi90** [comment]
> I think we have to make sure that if this PR lands before that API uplift work in other repos and corresponding changes in this repo (#693), we need to update those references most probably in #693 after rebasing.

I can parameterize the versions in this PR and then wait for the uplift to be ready before merging. The reason I now hard-coded `v1alpha3` was to be able to run the tests (they are broken if I use `CAPI_VERSION` since that variable is already set to `v1alpha4`). But now that I know that the tests are green I can put the variable back and wait for the uplift. Do you think this would be ok @furkatgofurov7 ?

**Comment by Xenwar** [comment]
> > I think we have to make sure that if this PR lands before that API uplift work in other repos and corresponding changes in this repo (#693), we need to update those references most probably in #693 after rebasing.
> 
> I can parameterize the versions in this PR and then wait for the uplift to be ready before merging. The reason I now hard-coded `v1alpha3` was to be able to run the tests (they are broken if I use `CAPI_VERSION` since that variable is already set to `v1alpha4`). But now that I know that the tests are green I can put the variable back and wait for the uplift. Do you think this would be ok @furkatgofurov7 ?

+1 for temporarily hard coding. Then creating a new PR for parameterize them.

**Comment by lentzi90** [comment]
> +1 for temporarily hard coding. Then creating a new PR for parameterize them.

If this is OK then we can merge it directly. :slightly_smiling_face: It should be simple to parameterize later, just find/replace. :+1: 
But if the uplift is very close to ready then I think this can wait so we don't cause extra work.

**Comment by Xenwar** [comment]
> > +1 for temporarily hard coding. Then creating a new PR for parameterize them.
> 
> If this is OK then we can merge it directly. 🙂 It should be simple to parameterize later, just find/replace. 👍
> But if the uplift is very close to ready then I think this can wait so we don't cause extra work.

I would go with merging it. In the worst case scenario, we just just create another hardcode value with the version after the uplift.

**Comment by furkatgofurov7** [comment]
> But now that I know that the tests are green I can put the variable back and wait for the uplift. Do you think this would be ok @furkatgofurov7 ?

I was just referring that if we take this in now, we have to do small change on #693 regarding the API version, which is fine and not a blocker for this to be merged. Changes lgtm.

**Comment by lentzi90** [comment]
Thanks @furkatgofurov7 ! Can you add `/lgtm` then? :wink:

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (kashifest) [inline vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:10]:
```suggestion
      name: {{ CLUSTER_NAME }}
```

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (kashifest) [inline vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:223]: @smoshiur1237 I hope you are also parameterizing these v1alpha3s in your PR

**Review by kashifest — COMMENTED** [review]
lgtm, small nit. Good job

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml]: Thanks, fixed!

Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml]: Thanks, fixed!

File: vm-setup/roles/v1aX_integration_test/filter_plugins/k8s_fields.py
Comment (mboukhalfa) [inline vm-setup/roles/v1aX_integration_test/filter_plugins/k8s_fields.py:18]: k8s `apply` https://docs.ansible.com/ansible/latest/collections/community/kubernetes/k8s_module.html#parameter-apply supposed to ignore properties that are automatically generated. Does not it help with those fields ?

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:8]: It won't work until we have support for v1alpha4. `CAPI_VERSION` is already set to v1alpha4 so we must wait for the uplift until we can start using this variable. :slightly_frowning_face:  See discussion above.

File: vm-setup/roles/v1aX_integration_test/filter_plugins/k8s_fields.py
Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/filter_plugins/k8s_fields.py:18]: Interesting! Will check, thanks!

Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/filter_plugins/k8s_fields.py:18]: Unfortunately it doesn't work. :slightly_frowning_face: I guess it ignores the auto generated properties when doing the comparison, but not when doing the actual update. This is how I tested it:

```yaml
- name: Create deployment
      kubernetes.core.k8s:
        api_version: apps/v1
        kind: Deployment
        name: test
        namespace: default
        resource_definition:
          metadata:
            labels:
              app: test
          spec:
            replicas: 1
            selector:
              matchLabels:
                app: test
            template:
              metadata:
                labels:
                  app: test
              spec:
                containers:
                  - name: nginx
                    image: nginx

    - name: Get deployment
      kubernetes.core.k8s_info:
        api_version: apps/v1
        kind: Deployment
        name: test
        namespace: default
      register: deployment

    - debug:
        msg: "{{ deployment }}"

    - name: Scale deployment
      kubernetes.core.k8s_scale:
        api_version: apps/v1
        kind: Deployment
        name: test
        namespace: default
        replicas: 2

    - name: Restore deployment
      kubernetes.core.k8s:
        api_version: apps/v1
        kind: Deployment
        name: test
        namespace: default
        apply: yes
        resource_definition: "{{ deployment.resources[0] }}"
```

In the last step there is always a failure due to status 409 (conflict). I believe this is because the resource version has changed. So the server knows that we are trying to update an object that we haven't seen the latest version of yet.

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (furkatgofurov7) [inline vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:18]: In general, are those scale operations working as intended? Because we have seen it silently ignoring the intended change and setting the default value for `maxSurge` field in node reuse tests which resulted in hiding the bug. 

cc @mboukhalfa

**Review by furkatgofurov7 — COMMENTED** [review]
Overall looks good to me.

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (mboukhalfa) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:18]: This should work because in node_reuse I was using `apply: yes ` which ignores those fields.

Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:18]: Yes scaling has worked well here. :slightly_smiling_face:

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (fmuyassarov) [inline vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:210]: Maybe we could add a loop to wait for the pods in the capi-system to be available ?

**Review by fmuyassarov — COMMENTED** [review]
Really good work!

File: vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml
Comment (lentzi90) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:210]: Do you know if that will be enough? I saw that there is also some things in `capi-webhook-system`, `capi-kubeadm-control-plane-system` and `capi-kubeadm-bootstrap-system`. Since I'm not really sure what we would have to wait for I just left it as a pause for now, but I can certainly add a wait for some/all of these deployments/pods if you know which are important! :slightly_smiling_face:

Comment (fmuyassarov) [reply vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml:210]: Ah true, there is also webhook namespace. Maybe we could take the namespaces that starts with `capi*`. But we could it in a separate PR, this should not be a blocker.

**Comment by lentzi90** [comment]
/hold
Wait with this until https://github.com/metal3-io/metal3-dev-env/pull/721 is done

**Comment by jan-est** [comment]
LGTM

**Pull request merged**
